samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+ */
+static void broker_init(void)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo *addr;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM
+	};
+
+	err = getaddrinfo(ctx.url, NULL, &hints, &result);
+	if (err) {
+		LOG_ERR("ERROR: getaddrinfo failed %d", err);
+
+		return;
+	}
+
+	addr = result;
+	err = -ENOENT;
+
+	/* Look for address of the broker. */
+	while (addr != NULL) {
+		/* IPv4 Address. */
+		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+			struct sockaddr_in *broker4 =
+				((struct sockaddr_in *)&broker);
+			char ipv4_addr[NET_IPV4_ADDR_LEN];
+
+			broker4->sin_addr.s_addr =
+				((struct sockaddr_in *)addr->ai_addr)
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(ctx.port);
+
+			inet_ntop(AF_INET, &broker4->sin_addr.s_addr,
+				  ipv4_addr, sizeof(ipv4_addr));
+			LOG_DBG("IPv4 Address found %s", ipv4_addr);
+
+			break;
+		}
+		LOG_DBG("ai_addrlen = %u should be %u or %u",
+			(unsigned int)addr->ai_addrlen,
+			(unsigned int)sizeof(struct sockaddr_in),
+			(unsigned int)sizeof(struct sockaddr_in6));
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+/**@brief Initialize the MQTT client structure
+ */
+static void client_init(void)
+{
+	/* Init MQTT client */
+	mqtt_client_init(&client);
+
+	/* Init MQTT broker */
+	broker_init();
+
+	/* MQTT client configuration */
+	client.broker = &broker;
+	client.evt_cb = mqtt_evt_handler;
+	client.client_id.utf8 = ctx.cid;
+	client.client_id.size = strlen(ctx.cid);
+	client.password = NULL;
+	client.user_name = NULL;
+	client.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client.rx_buf = rx_buffer;
+	client.rx_buf_size = sizeof(rx_buffer);
+	client.tx_buf = tx_buffer;
+	client.tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	if (ctx.sec_transport == true) {
+		struct mqtt_sec_config *tls_config;
+
+		tls_config = &(client.transport).tls.config;
+		tls_config->peer_verify = TLS_PEER_VERIFY_REQUIRED;
+		tls_config->cipher_list = NULL;
+		tls_config->cipher_count = 0;
+		tls_config->sec_tag_count = 1;
+		tls_config->sec_tag_list = (int *)&ctx.sec_tag;
+		tls_config->hostname = ctx.url;
+		client.transport.type = MQTT_TRANSPORT_SECURE;
+	} else {
+		client.transport.type = MQTT_TRANSPORT_NON_SECURE;
+	}
+}
+
+/**@brief Initialize the file descriptor structure used by poll.
+ */
+static int fds_init(struct mqtt_client *c)
+{
+	if (c->transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		fds.fd = c->transport.tcp.sock;
+	} else {
+#if defined(CONFIG_MQTT_LIB_TLS)
+		fds.fd = c->transport.tls.sock;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	fds.events = POLLIN;
+
+	return 0;
+}
+
+static int do_mqtt_connect(void)
+{
+	int err = -EINVAL;
+
+	if (ctx.connected) {
+		return -EINPROGRESS;
+	}
+
+	client_init();
+
+	/* Start MQTT listening thread */
+	mqtt_thread_id = k_thread_create(&mqtt_thread,
```

Comment:
The thread should be spawned once, not at each connect call. Should be moved to `slm_at_mqtt_init` (or created statically with K_THREAD_DEFINE macro, https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#spawning-a-thread)

Reply:
Now the thread is statically defined in K_THREAD_DEFINE.


K_THREAD_DEFINE(mqtt_thread, K_THREAD_STACK_SIZEOF(mqtt_thread_stack),
		mqtt_thread_fn, NULL, NULL, NULL,
		THREAD_PRIORITY, 0, K_NO_WAIT);